### PR TITLE
Feature: New milestone stage and solidification pipeline

### PR DESCRIPTION
**Pull request by DyrellC**
# Description of change
Merges the new milestone solidification logic. This includes the addition of the `MilestoneStage` in the `TransactionProcessingPipeline` for handling transactions immediately on receiving them, as well as the consolidation of the `LatestMilestoneTracker`, `MilestoneSolidifier` and `LatestSolidMilestoneTracker` into one efficient `MilesotneSolidifier` class. These changes streamline the milestone solidification process and no longer requires continuous pulling of transactions from the db, reducing the memory consumption and general CPU usage of the node. 

## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested
These changes have been put through every test at our disposal: 
- Synchronisation tests
- Internal network testing 
- Regression testing 
- Jmeter bench-marking 
- Spam load testing

## Change checklist
- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java [hunk 1]
```diff
@@ -21,20 +24,25 @@
...
+    private MilestoneService milestoneService;
+    private Hash cooAddress;
 
     /**
      * Creates a new {@link ReceivedStage}.
-     *
+     * 
      * @param tangle           The {@link Tangle} database used to store/update the transaction
-     * @param txSolidifier      The {@link TransactionSolidifier} used to store/update the transaction
+     * @param txSolidifier      The {@link TransactionValidator} used to store/update the transaction
```
Comment (acha-bill) [inline src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java:34]: huh?

File: src/test/java/com/iota/iri/network/pipeline/TransactionProcessingPipelineTest.java [hunk 2]
```diff
@@ -180,11 +194,68 @@ public void processingAValidNewTransactionFlowsThroughTheEntirePipeline() throws
         Mockito.verify(broadcastStage).process(Mockito.any());
     }
 
+    @Test
+    public void processingAValidMilestone() throws InterruptedException {
```
Comment (acha-bill) [inline src/test/java/com/iota/iri/network/pipeline/TransactionProcessingPipelineTest.java:198]: Nice.

File: src/main/java/com/iota/iri/network/pipeline/MilestoneStage.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            //If the milestone index is below the latest snapshot initial index, then abort the process
+            //Exempts index 0, as milestone objects don't require both transactions to hold the index
+            if (payload.getMilestoneIndex() < snapshotProvider.getLatestSnapshot().getInitialIndex() &&
+                    payload.getMilestoneIndex() != 0) {
+              return abort(ctx);
+            }
+
+            TransactionViewModel milestone = payload.getMilestoneTransaction();
+            int newMilestoneIndex = payload.getMilestoneIndex();
+            boolean isTail = (milestone.getCurrentIndex() == 0);
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply src/main/java/com/iota/iri/network/pipeline/MilestoneStage.java:65]: I suppose not huh?

File: src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java (hunk 1 above)
Comment (DyrellC) [reply src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java:34]: Must have happened during the conflict resolution while merging dev in, i'll touch that up.

**Pull request merged**